cl_sii/rtc/xml_utils.py

```diff
@@ -27,3 +34,67 @@ def _get_signature(self, root: Any) -> object:
             return root
         else:
             return self._find(root, "Signature", anywhere=False)
+
+
+###############################################################################
+# functions
+###############################################################################
+
+def verify_aec_signature(
+    aec_xml_doc: xml_utils.XmlElement,
+    aec_xml: AecXml
+) -> Optional[bool]:
+    """
+    Verify signature of AEC XML document ``aec_xml_doc``.
+
+    :param aec_xml_doc: An AEC XML document, as returned by ``xml_utils.parse_untrusted_xml()``.
+    :param aec_xml: An instance of ``data_models_aec.AecXml`` with the data in the "cesión"'s
+        AEC XML document parsed from `aec_xml_doc` by ``parse_aec.parse_aec_xml``.
+    :raises ValueError: If the attribute `signature_x509_cert_der` of the AecXml is None.
+    :raises Exception: on unrecoverable errors
+    """
+    signature_verified: Optional[bool]
+    signature_x509_cert: Optional[crypto_utils.X509Cert] = None
+
+    if aec_xml.signature_x509_cert_der is None:
+        raise ValueError("Field 'signature_x509_cert_der' can not be None.")
+
+    try:
+        signature_x509_cert = crypto_utils.load_der_x509_cert(
+            aec_xml.signature_x509_cert_der,
+        )
+    except ValueError:
+        signature_verified = None
+
+    try:
+        assert signature_x509_cert is not None
+        aec_xml_verifier = AecXMLVerifier()
+        xml_utils.verify_xml_signature(
+            aec_xml_doc,
+            trusted_x509_cert=signature_x509_cert,
+            xml_verifier=aec_xml_verifier,
+            xml_verifier_supports_multiple_signatures=True,
+        )
+    except xml_utils.XmlSignatureUnverified:
+        signature_verified = False
+        logger.debug("AEC's digital signature did not verify")
+    except xml_utils.XmlSignatureInvalid:
+        signature_verified = False
+        logger.debug("AEC's digital signature is invalid")
+    except AssertionError:
+        signature_verified = None
+        logger.debug(
+            "The X.509 certificate could not be loaded from AEC's digital "
+            "signature's DER-encoded X.509 certificate."
+        )
```

Comment:
Assertions are for debugging and testing purposes, and shouldn't be used for control flow, as they're not executed when optimization is enabled.

**Example**:

```sh
PYTHONOPTIMIZE=true make test
```

```
[…]

======================================================================
FAIL: test_ok_for_bad_certificate_value (tests.test_rtc_xml_utils.FunctionVerifyAecSignatureTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jtrh/projects/lib-cl-sii-python/vcs/tests/test_rtc_xml_utils.py", line 108, in test_ok_for_bad_certificate_value
    self.assertIsNone(is_signature_verified)
AssertionError: False is not None

[…]
```

**Related documentation**:

- https://docs.python.org/3.8/reference/simple_stmts.html#assert
- https://docs.python.org/3.8/using/cmdline.html#cmdoption-o

Reply:
> they're not executed when optimization is enabled